packages/aws-cdk/lib/api/hotswap/monitor/logs-monitor.ts

```diff
@@ -0,0 +1,192 @@
+import * as util from 'util';
+import * as colors from 'colors/safe';
+import { error } from '../../../logging';
+import { ISDK } from '../../aws-auth';
+
+// how often we should read events from CloudWatchLogs
+const SLEEP = 2_000;
+
+/**
+ * Represents a CloudWatch Log Event that will be
+ * printed to the terminal
+ */
+export interface CloudWatchLogEvent {
+  readonly message: string;
+  readonly logGroup: string;
+  readonly timestamp: Date;
+}
+
+/**
+ * Options for creating a CloudWatchLogEventMonitor
+ */
+export interface CloudWatchLogEventMonitorOptions {
+  /**
+   * The time that watch was first triggered.
+   * CloudWatch logs will be filtered using this start time
+   *
+   * @default - the current time
+   */
+  readonly hotswapTime?: Date;
+
+  /**
+   * The EventPrinter to use to print out CloudWatch events
+   *
+   * @default - a default printer will be created
+   */
+  readonly printer?: IEventPrinter;
+}
+
+export class CloudWatchLogEventMonitor {
+  /**
+   * Determines which events not to display
+   */
+  private startTime: number;
+
+  private logGroups = new Set<string>();
+  private sdk?: ISDK;
+
+  /**
+   * The event printer that controls printing out
+   * CloudWatchLog Events
+   */
+  public printer: IEventPrinter;
+
+  constructor(options: CloudWatchLogEventMonitorOptions = {}) {
+    this.startTime = options.hotswapTime?.getTime() ?? Date.now();
+    this.printer = options.printer ?? new EventPrinter({
+      stream: process.stderr,
+    });
+  }
+
+  public start(): CloudWatchLogEventMonitor {
+    // call tick every x seconds
+    setInterval(() => void (this.tick()), SLEEP);
+    return this;
+  }
+
+  public addSDK(sdk: ISDK): void {
+    this.sdk = sdk;
+  }
+
+  /**
+   * Adds a CloudWatch log group to read log events from
+   */
+  public addLogGroups(logGroups: string[]): void {
+    logGroups.forEach(group => {
+      this.logGroups.add(group);
+    });
+  }
+
+  private async tick(): Promise<void> {
+    try {
+      const events = await this.readNewEvents();
+      const flatEvents: CloudWatchLogEvent[] = Array.prototype.concat.apply([], events);
+      flatEvents.forEach(event => {
+        this.printer.print(event);
+      });
+      // don't update the startTime until we've processed events
+      // we may miss some events if they come in after we've read the
+      // events, but before we update the startTime
+      if (flatEvents.length > 0) {
+        this.startTime = Date.now();
+      }
+
+    } catch (e) {
+      error('Error occurred while monitoring logs: %s', e);
+    }
+  }
+
+  /**
+   * Reads all new log events from a set of CloudWatch Log Groups
+   * in parallel
+   */
+  private async readNewEvents(): Promise<Array<CloudWatchLogEvent[]>> {
```

Comment:
Let's make it obvious this returns a nested array:

```suggestion
  private async readNewEvents(): Promise<Array<Array<CloudWatchLogEvent>>> {
```